modules/network/src/test/java/org/apache/ignite/internal/network/serialization/MarshallableTest.java

```diff
@@ -0,0 +1,173 @@
+     */
+    @Test
+    public void testMarshallable() {
+        TestMessagesFactory msgFactory = new TestMessagesFactory();
+
+        ChannelHandlerContext ctx = Mockito.mock(ChannelHandlerContext.class);
+
+        var channel = new EmbeddedChannel();
+
+        Mockito.doReturn(channel).when(ctx).channel();
+
+        ClassDescriptorFactoryContext descriptorContext = new ClassDescriptorFactoryContext();
+        ClassDescriptorFactory factory = new ClassDescriptorFactory(descriptorContext);
+
+        // Creates descriptor for SimpleSerializableObject
+        ClassDescriptor descriptor = factory.create(SimpleSerializableObject.class);
+
+        // Stub implementation of the serializer, uses standard JDK serializable serialization to actually marshall an object
+        UserObjectSerializer userObjectSerializer = new UserObjectSerializer() {
+            @Override
+            public <T> T read(Map<Integer, ClassDescriptor> descriptor, byte[] array) {
+                try (ByteArrayInputStream bais = new ByteArrayInputStream(array); ObjectInputStream ois = new ObjectInputStream(bais)) {
+                    return (T) ois.readObject();
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            @Override
+            public <T> SerializationResult write(T object) {
+                try (ByteArrayOutputStream baos = new ByteArrayOutputStream(); ObjectOutputStream oos = new ObjectOutputStream(baos)) {
+                    oos.writeObject(object);
+                    oos.close();
+                    return new SerializationResult(baos.toByteArray(), Collections.singletonList(descriptor.descriptorId()));
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            @Override
+            public ClassDescriptor getClassDescriptor(int typeDescriptorId) {
+                return descriptorContext.getDescriptor(typeDescriptorId);
+            }
+
+            @Override
+            public ClassDescriptor getClassDescriptor(String typeName) {
+                assertEquals(descriptor.className(), typeName);
+
+                return descriptor;
+            }
+
+            @Override
+            public boolean isBuiltin(int typeDescriptorId) {
+                return ClassDescriptorFactoryContext.isBuiltin(typeDescriptorId);
+            }
+        };
+
+        var serializationService = new SerializationService(registry, userObjectSerializer);
+        var perSessionSerializationService = new PerSessionSerializationService(serializationService);
+        final var decoder = new InboundDecoder(perSessionSerializationService);
+
+        // List that holds decoded object
+        final var list = new ArrayList<>();
+
+        var writer = new DirectMessageWriter(perSessionSerializationService, ConnectionManager.DIRECT_PROTOCOL_VERSION);
```

Comment:
I suggest extracting different stages of the test to methods of their own, so that the high-level structure of the test (the script in large blocks) becomes easily visible. For example, methods for writing a message, for reading a message, and so on could be extracted.